src/datasources/accounts/accounts.datasource.ts

```diff
@@ -62,6 +64,76 @@ export class AccountsDatasource implements IAccountsDatasource {
   }
 
   async getDataTypes(): Promise<AccountDataType[]> {
+    // TODO: add caching with clearing mechanism.
     return this.sql<[AccountDataType]>`SELECT * FROM account_data_types`;
   }
+
+  async getAccountDataSettings(
+    address: `0x${string}`,
+  ): Promise<AccountDataSetting[]> {
+    const account = await this.getAccount(address);
+    return this.sql<[AccountDataSetting]>`
+      SELECT * FROM account_data_settings WHERE account_id = ${account.id}
+    `;
+  }
+
+  /**
+   * Adds or updates the existing account data settings for a given address/account.
+   * Requirements:
+   * - The account must exist.
+   * - The data type must exist.
+   * - The data type must be active.
+   *
+   * @param address - account address.
+   * @param upsertAccountDataSettings {@link UpsertAccountDataSettingsDto} object.
+   * @returns {Array<AccountDataSetting>} inserted account data settings.
+   */
+  async upsertAccountDataSettings(
+    address: `0x${string}`,
+    upsertAccountDataSettings: UpsertAccountDataSettingsDto,
+  ): Promise<AccountDataSetting[]> {
+    const { accountDataSettings } = upsertAccountDataSettings;
+    await this.checkDataTypes(accountDataSettings);
+    const account = await this.getAccount(address);
+    await this.sql.begin(async (sql) => {
+      return Promise.all(
+        accountDataSettings.map(async (accountDataSetting) => {
+          return sql`
+            INSERT INTO account_data_settings (account_id, account_data_type_id, enabled)
+            VALUES (${account.id}, ${accountDataSetting.id}, ${accountDataSetting.enabled})
+            ON CONFLICT (account_id, account_data_type_id) DO UPDATE SET enabled = EXCLUDED.enabled
+          `.catch((e) => {
+            throw new UnprocessableEntityException(
+              `Error updating data settings: ${asError(e).message}`,
+            );
+          });
+        }),
+      );
+    });
```

Comment:
Good point! I've included the final selection in the transaction in [7840a53](https://github.com/safe-global/safe-client-gateway/pull/1715/commits/7840a53a7fbd14e94d4b9428a7eaaec90bc7c712). I needed to replace the call to the class function `getAccountDataSettings` with the SQL query:

```
return sql<[AccountDataSetting]>`
        SELECT * FROM account_data_settings WHERE account_id = ${account.id}`;
```

This is due to the fact the function `getAccountDataSettings` uses `this.sql` instead of the `sql` instance provided inside the transaction. Otherwise, the result would not include the rows inserted inside the transaction.

This also has a positive side effect, as before by calling `getAccountDataSettings` it was executing an unnecessary `getAccount(address)` call.